tao_compiler/mlir/disc/transforms/disc_dot_merge.cc

```diff
@@ -0,0 +1,557 @@
+      SmallVector<Value, 4> end_values;
+      end_values.push_back(builder.create<arith::ConstantIndexOp>(loc, i + 1));
+      for (int64_t j = 1; j < result_rank; j++) {
+        end_values.push_back(
+            builder.create<tensor::DimOp>(loc, batched_dot, j));
+      }
+      auto index_ty = builder.getIndexType();
+      auto start_indices = builder.create<tensor::FromElementsOp>(
+          loc,
+          RankedTensorType::get({static_cast<int64_t>(begin_values.size())},
+                                index_ty),
+          begin_values);
+      auto end_indices = builder.create<tensor::FromElementsOp>(
+          loc,
+          RankedTensorType::get({static_cast<int64_t>(end_values.size())},
+                                index_ty),
+          end_values);
+      auto stride_indices = builder.create<tensor::FromElementsOp>(
+          loc,
+          RankedTensorType::get({static_cast<int64_t>(stride_values.size())},
+                                index_ty),
+          stride_values);
+      SmallVector<int64_t, 4> slice_shapes(result_rank,
+                                           ShapedType::kDynamicSize);
+      slice_shapes[0] = 1;
+      for (int64_t i = 1; i < result_rank; i++) {
+        slice_shapes[i] = orig_dot_type.getDimSize(i - 1);
+      }
+      auto slice_type =
+          RankedTensorType::get(slice_shapes, orig_dot_type.getElementType());
+      auto dyn_slice = builder.create<mhlo::RealDynamicSliceOp>(
+          loc, slice_type, batched_dot, start_indices, end_indices,
+          stride_indices);
+      SmallVector<Value, 4> reshape_shape_values;
+      for (int64_t j = 1; j < slice_type.getRank(); j++) {
+        reshape_shape_values.push_back(
+            builder.create<tensor::DimOp>(loc, dyn_slice, j));
+      }
+      auto reshape_shape = builder.create<tensor::FromElementsOp>(
+          loc,
+          RankedTensorType::get(
+              {static_cast<int64_t>(reshape_shape_values.size())}, index_ty),
+          reshape_shape_values);
+      auto dyn_reshape = builder.create<mhlo::DynamicReshapeOp>(
+          loc, orig_dot_type, dyn_slice, reshape_shape);
+      op->replaceAllUsesWith(dyn_reshape);
+    }
+  }
+
+  // No longer need the original dot ops.
+  for (int64_t i = 0; i < ops.size(); i++) {
+    ops[i]->erase();
+  }
+
+  return true;
+}
+
+bool DotBatchingConverter::batchingDots() {
+  // Dot ops in `equal_shape_dot_map_` can be batched together if the batching
+  // does not introduce cycle.
+
+  // Form cycle detector.
+  std::vector<Operation*> func_op_list;
+  DenseMap<Operation*, int64_t> func_op_to_id;
+  func_.walk([&](Operation* op) {
+    func_op_to_id.try_emplace(op, func_op_list.size());
+    func_op_list.push_back(op);
+  });
+  GraphCycles cycle_detector(func_op_list.size());
+  for (int node_id = 0; node_id < func_op_list.size(); ++node_id) {
+    Operation* op = func_op_list[node_id];
+    for (Value operand : op->getOperands()) {
```

Comment:
op having regions could also consume values outside and thus we need to take these implicit dependencies into consideration as well . please check the logic in the fusion pass to find more details.

Reply:
Thanks for the comment. I will check that.

Reply:
Done.